elastic/beats
# Move around librpm calls, add `setreuid()` safety net

## Proposed commit message

This is an odd one. It turns out that the BDB backend used by older `rpm` installs is...really easy to corrupt, and also prone to a variety of crashes should you get too aggressive with spinning up multiple threads to access the RPM backend. The only way I've found to actually solve this in auditbeat is to use `setreuid()` to drop out of root before we make any librpm calls; this prevents librpm from touching any lock files or opening any database files in write mode, which means we can't corrupt the DB if auditbeat stops/crashes, and it also seems to reduce the odds of certain crashes originating from within the guts of librpm/libdb, as we can't perform any of the integrity checking/cleanup operations that we would if the database state files were writable. 

The resulting `package.rpm_drop_to_uid` is not enabled by default, as we need a UID to drop to, and also setting one thread to a different user is (according to the man page for `setuid`) not POSIX-compliant. This is meant as a safety net for users on older linux distros that haven't migrated to sqlite for the RPM backend.

This also moves around some of the librpm calls, we now call `rpmReadConfigFiles()` earlier in the startup process, and we also call `rpmtsSetRootDir()`. This mostly prevents a few confusing error messages when librpm fails, but I woudn't be surprised if it was causing some more subtle bugs as well.

## Checklist



## Disruptive User Impact

This will not alter any default behavior. 

## How to test

You can test this on any linux distro to confirm that no existing behavior has changed. 

However, to test what this is actually meant to prevent, you'll need to spin up a Linux box that uses rpm/BDB. This should be RHEL8 or any RHEL8-derivatives (Rocky8, etc). Then:

1)  Set run `go test -c` in the `package` directory
2) run multiple instances of the `TestWithSuid` test as root: `sudo ./package.test -test.v -test.run TestWithSuid` at once: `for i in {1..30}; do ./package.test -test.v -test.run TestWithSuid & done`
3) Make sure nothing crashes
4) If librpm segfaults anyway, make sure the DB isnt't corrupted: `sudo rpm -qa`.
